core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCache.java

```diff
@@ -60,38 +61,35 @@ protected ChunkCache(final ChunkManager chunkManager) {
      * opened right when fetching from cache happens even if the actual value is removed from the cache,
      * the InputStream will still contain the data.
      */
-    public InputStream getChunk(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+    public InputStream getChunk(final RemoteLogSegmentMetadata segmentMetadata,
                                 final SegmentManifest manifest,
                                 final int chunkId) throws StorageBackendException, IOException {
-        final Uuid id = remoteLogSegmentMetadata.remoteLogSegmentId().id();
+        final Uuid id = segmentMetadata.remoteLogSegmentId().id();
         final ChunkKey chunkKey = new ChunkKey(id, chunkId);
         final AtomicReference<InputStream> result = new AtomicReference<>();
         try {
-            return cache.asMap().compute(chunkKey, (key, val) -> CompletableFuture.supplyAsync(() -> {
-                if (val == null) {
+            final var cacheMap = cache.asMap();
+            cacheMap
+                .computeIfAbsent(chunkKey, key -> CompletableFuture.supplyAsync(() -> {
                     try {
-                        final InputStream chunk = chunkManager.getChunk(remoteLogSegmentMetadata, manifest, chunkId);
+                        final InputStream chunk = chunkManager.getChunk(segmentMetadata, manifest, chunkId);
                         final T t = this.cacheChunk(chunkKey, chunk);
                         result.getAndSet(cachedChunkToInputStream(t));
                         return t;
                     } catch (final StorageBackendException | IOException e) {
                         throw new CompletionException(e);
                     }
-                } else {
-                    try {
-                        final T cachedChunk = val.get();
-                        result.getAndSet(cachedChunkToInputStream(cachedChunk));
-                        return cachedChunk;
-                    } catch (final InterruptedException | ExecutionException e) {
-                        throw new CompletionException(e);
-                    }
-                }
-            }, executor))
-                    .thenApplyAsync(t -> result.get())
+                }, executor))
+                .get(GET_TIMEOUT_SEC, TimeUnit.SECONDS);
+            if (result.get() == null) {
+                return cacheMap.get(chunkKey)
+                    .thenApplyAsync(this::cachedChunkToInputStream)
                     .get(GET_TIMEOUT_SEC, TimeUnit.SECONDS);
+            } else {
+                return result.get();
+            }
```

Comment:
Can we avoid touching this altogether? It was specifically done like this to guarantee thread-safety + providing not a cached input stream but creating a new one on every get to the cache at the same time preserving the value of the cache to be InputStream. That brings some additional benefits under the hood also.

Reply:
I wouldn't mind. Unfortunately, with `compute` there's no metric computation as the action is ad-hoc, and cannot count as hit or missed [1].
If we feel strongly about it, then will have to define custom metrics to be handled on top of this implementation 

[1] https://github.com/ben-manes/caffeine/issues/689

Reply:
After offline discussion, we are trying a new approach by calling StatCounter manually as part of the `compute` method.